library/qs_source/src/std/diagnostics.qs

```diff
@@ -149,5 +149,25 @@ namespace Microsoft.Quantum.Diagnostics {
         areEqual
     }
 
-    export DumpMachine, DumpRegister, CheckZero, CheckAllZero, Fact, CheckOperationsAreEqual;
+    @Config(Unrestricted)
+    operation StartCountingOperation<'In, 'Out>(callable : 'In => 'Out) : Unit {
+        body intrinsic;
+    }
+
+    @Config(Unrestricted)
+    operation StopCountingOperation<'In, 'Out>(callable : 'In => 'Out) : Int {
+        body intrinsic;
+    }
+
+    @Config(Unrestricted)
+    operation StartCountingFunction<'In, 'Out>(callable : 'In -> 'Out) : Unit {
```

Comment:
Do we need distinct APIs for operations and functions, versus just having one Start/Stop pair for any callable?

Reply:
Yeah, because they are distinct types in Q#. A generic operation is `'In => 'Out` while a generic function is `'In -> 'Out`. There is no way to express something that is any generic callable (something something type classes grumble grumble). I could have one Start/Stop that takes generic `'T` but then you could pass anything and passing the wrong type would result in a runtime failure, which is not a good experience and is inconsistent with our goals of type safety for the language. See our old friend https://github.com/microsoft/qsharp/issues/481.